### PR TITLE
Updateable layout_provider/graph_layout

### LINKED ISSUE #11474: [BUG] graph_renderer.layout_provider changes via JS code are not displayed on plots
When the  graph_renderer.layout_provider  is changed from JS perspective (from user side), the changes are not displayed on any network plot. What I would like to do is to change the layout_provider (the position of nodes) from client’s side. Not from server. 
You can find the msg exchanged in discourse.bokeh.org [here ](https://discourse.bokeh.org/t/change-graphrender-position-from-client-side-js-callback/8127/5) 
```python
-----------------------------
below the sample code: 
----------------------------

#Bokeh_question.py
#-*- coding: utf-8 -*-
#-----------------------------------------------------------------------------
#>> Imports:
#-----------------------------------------------------------------------------

#ΝetworkΧ

import networkx as nx

#Bokeh library

from bokeh.io import output_file
from bokeh.plotting import figure, show
from bokeh.layouts import layout
from bokeh.models.graphs import from_networkx, NodesAndLinkedEdges, EdgesAndLinkedNodes

from bokeh.models import (
    ColumnDataSource,
    CustomJS,
    Circle,
    HoverTool,
    MultiLine,
    Range1d,
    ResetTool,
    TapTool,
    BoxSelectTool,
    Select,
    PanTool,
    WheelZoomTool,
    SaveTool,
    LabelSet,
)

#-----------------------------------------------------------------------------
#<< Imports.
#-----------------------------------------------------------------------------

#-----------------------------------------------------------------------------
#>> NetworX graph creation:
#-----------------------------------------------------------------------------

#define the graph's elements:
source = ['C1', 'C1', 'C1','C3','C2','C2','C2','C4','C5']
target = ['C3', 'C2','C4','C4','C3','C4','C1','C3', 'C3']
weight = [1,0.3,1,1,1,0.8, 0.4, 1, 1]

nodes_disc = {
    'C1': 'C1 description',
    'C2': 'C2 description',
    'C3': 'C3 description',
    'C4': 'C4 description',
    'C5': 'C5 description',
}

#create list of tuples with all necessary elements for graph
edges_list = list(zip(source, target, weight))

#create netwokx graph:
nx_graph = nx.DiGraph()
nx_graph.add_weighted_edges_from(edges_list)

#-----------------------------------------------------------------------------
#<< NetworX graph creation.
#-----------------------------------------------------------------------------

#-----------------------------------------------------------------------------
#>> Output HTML file:
#-----------------------------------------------------------------------------

#The figure will be rendered in a static HTML file called foo.html
output_file('FooF.html', title='Foo')

#-----------------------------------------------------------------------------
#<< Output HTML file.
#-----------------------------------------------------------------------------


#-----------------------------------------------------------------------------
#>> Figure:
#-----------------------------------------------------------------------------

plot = figure(
    plot_width=400,
    plot_height=350,
    x_range=Range1d(-3,5),
    y_range=Range1d(-3,5),
    tools=[],
)

plot.grid.grid_line_color = None

#plot title 
plot.title.text = "Foo"
plot.title.text_font = "times"

#Plot toolbar tools:
node_hover_tool = HoverTool()
plot_wheel_zoom = WheelZoomTool()
plot_pan_tool = PanTool()
plot.add_tools(
    plot_pan_tool,
    node_hover_tool,
    plot_wheel_zoom,
    BoxSelectTool(),
    ResetTool(),
    TapTool(),
    SaveTool(),
)

#active tools:
plot.toolbar.active_inspect = node_hover_tool
plot.toolbar.active_scroll = plot_wheel_zoom
plot.toolbar.active_drag = plot_pan_tool
plot.toolbar.active_tap = None
plot.toolbar.logo = None

#define some stitic nodes. e.g. Input or Output nodes:
node_initial_pos = {'C1':(0, 0), 'C2':(0, 3), 'C5':(3,0)}

##### graph renderer

#Different layout positions
bipartite_pos= nx.bipartite_layout(
    nx_graph,
    ['C1', 'C2'],
    align='vertical',
    scale=2,
    center=(0, 0),
    aspect_ratio=1.33,
)

circular_pos = nx.circular_layout(
    nx_graph,
    scale=2,
    center=(0, 0),
    dim=2,
)

kamada_kawai_pos = nx.kamada_kawai_layout(
    nx_graph,
    weight='weight',
    scale=2,
    center=(0, 0),
    dim=2,
)

planar_pos = nx.planar_layout(
    nx_graph,
    scale=2,
    center=(0, 0),
    dim=2,
)

random_pos = nx.random_layout(
    nx_graph,
    center=(0, 0),
    dim=2,
)

spring_pos = nx.spring_layout(
    nx_graph,
    scale=2,
    center=(0, 0),
    pos=node_initial_pos,
    fixed=['C1', 'C2', 'C5'],
    iterations=2,
    k=0.1,
)

pos_dict = planar_pos

graph_renderer = from_networkx(
    nx_graph,
    pos_dict,
)

##node sub-renderer

graph_renderer.node_renderer.glyph = Circle(size=20, fill_color="#2b83ba", line_width=1.5, fill_alpha=0.7)
graph_renderer.node_renderer.nonselection_glyph = Circle(size=10, fill_color='#DAF7A8')
graph_renderer.node_renderer.selection_glyph = Circle(size=22, fill_color='#E74C3C')
graph_renderer.node_renderer.hover_glyph = Circle(size=15, fill_color='#DAF7A6')

##edge sub-renderer
graph_renderer.edge_renderer.glyph = MultiLine(line_color="grey", line_alpha=0.8, line_width=1.25)
graph_renderer.edge_renderer.selection_glyph = MultiLine(line_color='#E74C3C', line_width=2)
graph_renderer.edge_renderer.nonselection_glyph = MultiLine(line_color='#DAF7A8', line_width=1)
graph_renderer.edge_renderer.hover_glyph = MultiLine(line_color='#DAF7A6', line_width=2) # #abdda4

#Hover policy
graph_renderer.selection_policy = NodesAndLinkedEdges()
graph_renderer.inspection_policy = EdgesAndLinkedNodes()

#append graph renderer to plot figure
plot.renderers.append(graph_renderer)

#Adding concept names to Figure:

node_names = list(graph_renderer.layout_provider.graph_layout.keys())
_xs, _ys = map(list, zip(*graph_renderer.layout_provider.graph_layout.values()))
node_labels_source = ColumnDataSource(
    data=dict(
        xs=_xs,
        ys=_ys,
        names=node_names,
    )
)

#Node labels layout:

node_labels = LabelSet(
    x='xs',
    y='ys',
    text='names',
    x_offset=0,
    y_offset=-5,
    source=node_labels_source,
    render_mode='canvas',
    text_font = 'times',
    text_font_size='9pt',
    text_font_style='normal',
    text_color='black',
    text_align='center',
    level='glyph',
     angle=0,
)

plot.add_layout(node_labels)

#>> Widgets related to FCM display layout:
##########################################

_layout_options = [
    'spring_layout',
    'bipartite_layout',
    'circular_layout',
    'kamada_kawai_layout',
    'planar_layout',
    'random_layout',
]

nodes_display_layout_select = Select(
    title="Display layout:",
    value="planar_layout",
    options=_layout_options,
    width=300,
    height=50
)

nodes_display_layout_select_args = dict(
    graph_renderer=graph_renderer,
    plot=plot,
    bipartite_pos=bipartite_pos,
    circular_pos=circular_pos,
    kamada_kawai_pos=kamada_kawai_pos,
    planar_pos=planar_pos,
    random_pos=random_pos,
    spring_pos=spring_pos,
)

nodes_display_layout_select_str = """
    graph_renderer.layout_provider.graph_layout = spring_pos;  
    graph_renderer.change.emit();
    plot.reset.change();
"""

nodes_display_layout_select_callback = CustomJS(
    args=nodes_display_layout_select_args,
    code=nodes_display_layout_select_str,
)

nodes_display_layout_select.js_on_change("value", nodes_display_layout_select_callback)

#<< Widgets related to FCM display layout. 
##########################################

##Final Layout:

html_layout = layout(
    [
    [plot],
    [nodes_display_layout_select ],
    ]
)

#Show HTML Layout:
show(html_layout)

```

### PR DESCRIPTION
**Pull request by g-parki**
- [x] issues: fixes #11474
- [x] tests added / passed

Edit after overhauling:

This PR is to fix the original issue of the graph not updating after the `layout_provider` was updated, and also add transform classes for edge/node coordinates:

In `layout_provider.ts`:
 - Base class `GraphCoordinates` extends `CoordinateTransform`, adds `layout` property
 - Classes `NodeCoordinates` and `EdgeCoordinates` extend `GraphCoordinates`, adds `_v_compute()` method
 - `layout_provider` has getters for `node_coordinates` and `edge_coordinates`

In `coordinate_transform.ts`:
 - CoordinateTransform component types broadened from `Float64Array` to `Arrayable<number> | Arrayable<number>[]`

Screen capture of the graph updating:
graph_render

For reference, here's the script I used for prototyping (based on issue #11474):
```
import networkx as nx
from bokeh.plotting import figure, show, from_networkx
from bokeh.layouts import layout
from bokeh.models import ColumnDataSource, CustomJS, Select, LabelSet

def get_node_label_source(graph, pos):
    renderer = from_networkx(graph, pos)
    node_names = list(renderer.layout_provider.graph_layout.keys())
    _xs, _ys = map(list, zip(*renderer.layout_provider.graph_layout.values()))
    return ColumnDataSource(
        data=dict(
            xs=_xs,
            ys=_ys,
            names=node_names
        )
    )

#define the graph's elements:
source = ['C1', 'C1', 'C1','C3','C2','C2','C2','C4','C5']
target = ['C3', 'C2','C4','C4','C3','C4','C1','C3', 'C3']
weight = [1,0.3,1,1,1,0.8, 0.4, 1, 1]
edges_list = list(zip(source, target, weight))

nx_graph = nx.DiGraph()
nx_graph.add_weighted_edges_from(edges_list)
plot = figure()
node_initial_pos = {'C1':(0, 0), 'C2':(0, 3), 'C5':(3,0)}

_layout_options = ['circular_pos', 'kamada_kawai_pos', 'planar_pos']
circular_pos = nx.circular_layout(nx_graph,scale=2,center=(0, 0),dim=2,)
kamada_kawai_pos = nx.kamada_kawai_layout(nx_graph,weight='weight',scale=2,center=(0, 0),dim=2,)
planar_pos = nx.planar_layout(nx_graph,scale=2,center=(0, 0),dim=2)

initial_graph_renderer = from_networkx(nx_graph, planar_pos)
plot.renderers.append(initial_graph_renderer)

initial_node_labels = LabelSet(
    x='xs',
    y='ys',
    text='names',
    source=get_node_label_source(nx_graph, planar_pos),
)
plot.add_layout(initial_node_labels)

nodes_display_layout_select = Select(
    title="Display layout:",
    value="planar_layout",
    options=_layout_options,
    width=300,
    height=50
)

nodes_display_layout_select_args = dict(
    graph_renderer=initial_graph_renderer,
    plot=plot,
    positions = dict(
        circular_pos=dict(position=circular_pos, label_source= get_node_label_source(nx_graph, circular_pos)),
        kamada_kawai_pos=dict(position=kamada_kawai_pos, label_source= get_node_label_source(nx_graph, kamada_kawai_pos)),
        planar_pos=dict(position=planar_pos, label_source= get_node_label_source(nx_graph, planar_pos))
    ),
    node_labels= initial_node_labels
)

nodes_display_layout_select_str = """
    const selected_layout = positions[cb_obj.value];
    graph_renderer.layout_provider.graph_layout = selected_layout.position;
    node_labels.source= selected_layout.label_source;
"""
nodes_display_layout_select_callback = CustomJS(
    args=nodes_display_layout_select_args,
    code=nodes_display_layout_select_str,
)
nodes_display_layout_select.js_on_change("value", nodes_display_layout_select_callback)

html_layout = layout(plot,nodes_display_layout_select)
show(html_layout, browser="windows-default")
```

### CODE REVIEW

File: bokehjs/test/unit/models/renderers/graph_renderer.ts [hunk 1]
```diff
@@ -9,34 +9,54 @@ import {Plot} from "@bokehjs/models/plots"
...
+      7: [6, 3],
+    }
+
+    graph.layout_provider = layout_provider
+
+    const node_glyph = graph.node_renderer.glyph as XYGlyph
+    const x = node_glyph.x as any
+    const y = node_glyph.y as any
+    const express_x = x.expr as Expression
+    const express_y = y.expr as Expression
```
Comment (mattpap) [inline bokehjs/test/unit/models/renderers/graph_renderer.ts:58]: Much better would be:
```ts
import {assert} from "core/util/assert"
import {isExpr} from "core/vectorization"
assert(isExpr(node_glyph.x))
```
After the `assert(...)` the type of `x` will be an `Expr`.

File: bokehjs/test/unit/models/renderers/graph_renderer.ts [hunk 2]
```diff
@@ -9,34 +9,54 @@ import {Plot} from "@bokehjs/models/plots"
...
+const node_renderer = new GlyphRenderer({
+  glyph: new Circle({size: 10, fill_color: "red"}),
+  data_source: new ColumnDataSource({data: {index: [4, 5, 6, 7]}}),
+})
+const edge_renderer = new GlyphRenderer({
+  glyph: new MultiLine({line_width: 2, line_color: "gray"}),
+  data_source: new ColumnDataSource({data: {start: [4, 4, 5, 6], end: [5, 6, 6, 7]}}),
+})
+
+const graph = new GraphRenderer({layout_provider, node_renderer, edge_renderer})
```
Comment (mattpap) [inline bokehjs/test/unit/models/renderers/graph_renderer.ts:33]: This code needs to be wrapped in a function and fresh objects need to be produced for every test. This way we can make sure that tests can be run e.g. out-of-order, and there will be no mutable state creeping between tests.

**Comment by mattpap** [comment]
I would like a visual regression test to be added, similar to:
https://github.com/bokeh/bokeh/blob/d3fce050d3079541cc1ace704ce80c8e2ad199a1/bokehjs/test/integration/regressions.ts#L51-L66
Test plot area should be as small as possible, but allowing visual inspection, and the data simple, but covering the issue.

**Comment by g-parki** [comment]
In `graph_renderer` I noticed there was a TODO to upgrade node/edge coordinates from an anonymous class to a transform. I was already having a typecasting nightmare with my unit tests, so figured it's a good time for the upgrade. Here's the main summary:

- In `layout_provider.ts`:
  - Classes `NodeCoordinates` and `EdgeCoordinates` extend `CoordinateTransform`
  - `layout_provider` has non-generic getters for each
- In `coordinate_transform.ts`:
  -  CoordinateTransform type broadened from `Expression<{x: Float64Array, y: Float64Array}>` to `Expression<{x: CoordinateType, y: CoordinateType}>` where `CoordinateType = Arrayable<number> | Arrayable<number>[]`

**Comment by g-parki** [comment]
Running list still to do (assuming this direction is okay)...

1. Implement same models in Bokeh (currently only in JS). This was caught in ci/js-test.
2. Address failing macos tests.

File: bokehjs/test/unit/models/graphs/graph_hit_test_policy.ts [hunk 3]
```diff
@@ -15,20 +15,26 @@ import {ColumnarDataSource} from "@bokehjs/models/sources/columnar_data_source"
...
+  constructor(attrs?: Partial<LayoutProvider.Attrs>) {
+    super(attrs)
+  }
+
+  get_node_coordinates(graph_source: ColumnarDataSource): [Float64Array, Float64Array] {
     const n = graph_source.get_length() ?? 1
-    return [new Float64Array(n), new Float64Array(n)]
+    const x = new Float64Array(n)
+    const y = new Float64Array(n)
+    return [x, y]
```
Comment (g-parki) [inline bokehjs/test/unit/models/graphs/graph_hit_test_policy.ts:30]: Didn't mean to keep these changes. Will remove.

**Comment by g-parki** [comment]
@bryevdv sorry I didn't realize 2.4 is due tomorrow. I can bring this PR back to minimally fixing #11474 if the rest is too much to consider by then...

**Comment by bryevdv** [comment]
@g-parki my guess is that the release will slip until Monday, there some loose ends that still need cleaning up. But it's fairly easy now for us to follow on with small point releases so I would say there is no need to rush or hurry, whatever pace works for you is fine.

File: bokeh/models/graphs.py [hunk 4]
```diff
@@ -57,7 +62,13 @@ class LayoutProvider(Model):
...
     '''
 
-    pass
+    @property
+    def node_coordinates(self):
+        return NodeCoordinates(layout= self)
+
+    @property
+    def edge_coordinates(self):
+        return EdgeCoordinates(layout= self)
```
Comment (mattpap) [inline bokeh/models/graphs.py:71]: Whitespace after `=` should be removed.

Comment (mattpap) [reply bokeh/models/graphs.py:71]: New code should have explicit (return) type annotations (even if trivial).

File: bokeh/models/graphs.py [hunk 5]
```diff
@@ -78,6 +89,30 @@ class StaticLayoutProvider(LayoutProvider):
         }
     """)
 
+@abstract
+class GraphCoordinates(CoordinateTransform):
+    '''
+    Abstract class for coordinate transform expression obtained from `LayoutProvider`
```
Comment (mattpap) [inline bokeh/models/graphs.py:95]: Sphinx requires double backticks for code blocks.

File: bokehjs/src/lib/models/graphs/layout_provider.ts [hunk 6]
```diff
@@ -21,4 +22,75 @@ export abstract class LayoutProvider extends Model {
...
+    layout: p.Property<LayoutProvider>
+  }
+}
+
+export interface GraphCoordinates extends GraphCoordinates.Attrs {}
+
+export abstract class GraphCoordinates extends CoordinateTransform {
+  override properties: GraphCoordinates.Props
+
+  constructor(attrs?: Partial<GraphCoordinates.Attrs>){
```
Comment (mattpap) [inline bokehjs/src/lib/models/graphs/layout_provider.ts:47]: A single space is required before opening a block. There are multiple other instances to fix. I will add a lint rule for this.

**Comment by mattpap** [comment]
@g-parki, looks good. Please rebase and fix the trivia and this should be ready for merging.

**Comment by g-parki** [comment]
> @g-parki, looks good. Please rebase and fix the trivia and this should be ready for merging.

@mattpap thanks for reviewing. I rebased and incorporated your suggestions.

**Pull request merged**